doc/neps/nep-0033-array-creation-dispatch-with-array-function.rst

```diff
@@ -0,0 +1,184 @@
+introduction of an overriding mechanism called ``uarray`` by NEP-31 [3]_.
+
+The purpose of this NEP is to address that shortcoming in a simple and
+straighforward way: introduce a new ``like=`` keyword argument, similar to how
+the ``empty_like`` family of functions work. When array creation functions
+receive such an argument, they will trigger the ``__array_function__`` protocol,
+and call the downstream library's own array creation function implementation.
+The ``like=`` argument, as its own name suggests, shall be used solely for the
+purpose of identifying where to dispatch.  In contrast to the way
+``__array_function__`` has been used so far (the first argument identifies where
+to dispatch), and to avoid breaking NumPy's API with regards to array creation,
+the new ``like=`` keyword shall be used for the purpose of dispatching.
+
+Usage Guidance
+~~~~~~~~~~~~~~
+
+The new ``like=`` keyword is solely intended to identify the downstream library
+where to dispatch and the object is used only as reference, meaning that no
+modifications, copies or processing will be performed on that object.
+
+We expect that this functionality will be mostly useful to library developers,
+allowing them to create new arrays for internal usage based on arrays passed
+by the user, preventing unnecessary creation of NumPy arrays that will
+ultimately lead to an additional conversion into a downstream array type.
+
+Implementation
+--------------
+
+The implementation requires introducing a new ``like=`` keyword to all existing
+array creation functions of NumPy. As of the writing of this NEP, a complete
+list of array creation functions can be found in [4]_.
+
+This newly proposed keyword shall be removed by the ``__array_function__``
+mechanism from the keyword dictionary before dispatching. The purpose for this
+is twofold:
+
+1. The object will have no use in the downstream library's implementation; and
+2. Simplifies adoption of array creation by those libraries already opting-in
+   to implement the ``__array_function__`` protocol, thus removing the
+   requirement to explicitly opt-in for all array creation functions.
+
+Downstream libraries thus shall _NOT_ include the ``like=`` keyword to their
+array creation APIs, which is a NumPy-exclusive keyword.
+
+Function Dispatching
+~~~~~~~~~~~~~~~~~~~~
+
+There are two different cases to dispatch: Python functions, and C functions.
+To permit ``__array_function__`` dispatching, one possible implementation is to
+decorate Python functions with ``overrides.array_function_dispatch``, but C
+functions have a different requirement, which we shall describe shortly.
+
+The example below shows a suggestion on how the ``asarray`` could be decorated
+with ``overrides.array_function_dispatch``:
+
+.. code:: python
+
+    def _asarray_decorator(a, dtype=None, order=None, like=None):
+        if like is not None:
+            return (like,)
+        return (a,)
```

Comment:
I think you need to justify why this isn't
```
if like is not None:
    return (a, like)
return (a,)
```

Reply:
I'm puzzled why this isn't just:
```python
if like is not None:
    return (like,)
return ()
```
or perhaps just:
```python
return (like,)
```

My understanding is that `a` should not participate in dispatching, at least not if `like` is provided, because that would present potential backwards compatible issue.

Reply:
I agree with Stephan here. There was a reason `(a,)` was not used on `asarray` in the first place.

Reply:
Yes, @shoyer and @hameerabbasi are absolutely right. The reason for this is I started working on an example for another function and ended up mixing things. Also Stephan's solution is much more elegant, thanks for suggesting that and the fix is now in.